File: src/com/train/algorithm/list/SlidingWindow.java
package com.train.algorithm.list;

import java.util.HashMap;
import java.util.Map;

// 滑动窗口合集，通常与hashmap组合
public class SlidingWindow {
    //leetcode-1663
    public String getSmallestString(int n, int k) {
        char[] c = new char[n];
        StringBuilder str = new StringBuilder();
        for(int i = n - 1;i >= 0;i--){
            if(k - i > 26){
                c[i] = 'z';
                k -= 26;
            }else{
                if(k == 1){
                    c[i] = (char)(k + 96);
                    continue;
                }
                c[i] = (char)(k - i + 96);
                k = 1;
            }
        }
        for(int i = 0;i < n;i++){
            str.append(c[i]);
        }
        return str.toString();
    }

    //leetcode-904
    public int totalFruit(int[] tree) {
        Map<Integer,Integer> window = new HashMap<>();
        int count = 0,left = 0,right = 0,maxlen = 0;
        while(right < tree.length){
            if(!window.containsKey(tree[right])||window.get(tree[right])==0){
                count++;
            }
            window.put(tree[right],window.getOrDefault(tree[right],0) + 1);
            right++;
            if(count <= 2){
                maxlen = Math.max(maxlen,right-left);
            }
            while(count > 2){
                window.put(tree[left],window.getOrDefault(tree[left],0) - 1);
                if(window.get(tree[left]) == 0){
                    count--;
                }
                left++;
            }
        }
        return maxlen;
    }

    //leetcode-3
    public int lengthOfLongestSubstring(String s) {
        if(s.length() == 0){
            return 0;
        }
        Map<Character,Integer> map = new HashMap();
        int len = 0,left = 0;
        for(int i = 0;i < s.length();i++){
            if(map.containsKey(s.charAt(i))){
                left = Math.max(left,map.get(s.charAt(i)) + 1);
            }
            map.put(s.charAt(i),i);
            len = Math.max(len,i - left + 1);
        }
        return len;
    }
}

File: src/com/train/algorithm/list/implementInJS/SlidingWindow.js
// 滑动窗口
// leetcode-1663
var getSmallestString = function(n, k) {
	let res = Array(n).fill('a'),remain = k - n,i = n - 1;
	while(remain){
		if(remain > 25){
			remain -= 25;
			res[i] = 'z';
			i--;
		}else{
			res[i] = String.fromCharCode(97 + remain);
			remain = 0;
		}

	}
	return res.join('');
};

// leetcode-904
var totalFruit = function(fruits) {
	let l = 0,maxLen = 0,n = 0;
	let arr = [fruits[l]];

	for(let r = 0;r < fruits.length;r++){
		if(!arr.includes(fruits[r])){
			if(arr.length <= 1){
				arr[1] = fruits[r];
			}else{
				l = n;
				arr[0] = fruits[l];
				arr[1] = fruits[r];
			}
		}
		if(fruits[r] != fruits[n]){
			n = r;
		}
		maxLen = Math.max(maxLen,r - l + 1);
	}


	return maxLen;
};

// leetcode-3
var lengthOfLongestSubstring = function(s) {
	const o = new Set();
	const len = s.length;
	let right = -1,res = 0;
	for(let i = 0;i < len;i++){
		if(i != 0){
			o.delete(s.charAt(i - 1))
		}
		while(right + 1 < len && !o.has(s.charAt(right + 1))){
			o.add(s.charAt(right + 1));
			right++;
		}
		res = Math.max(res,right - i + 1);
	}
	return res;
};